pgxn/neon/relkind_cache.c

```diff
@@ -0,0 +1,334 @@
+/*-------------------------------------------------------------------------
+ *
+ * relkind_cache.c
+ *      Cache for marking unlogged relations
+ *
+ * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "neon_pgversioncompat.h"
+
+#include "pagestore_client.h"
+#include RELFILEINFO_HDR
+#include "storage/smgr.h"
+#include "storage/lwlock.h"
+#include "storage/ipc.h"
+#include "storage/shmem.h"
+#include "catalog/pg_tablespace_d.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+
+#if PG_VERSION_NUM >= 150000
+#include "miscadmin.h"
+#endif
+
+/*
+ * The main goal of this cache is to avoid calls of mdexists in neon_write,
+ * which is needed to distinguish unlogged relations.
+ *
+ * This hash is also used to mark relation during unlogged build.
+ * It has limited size, implementing eviction based on LRU algorithm.
+ * Relations involved in unlogged build are pinned in the cache (assuming that
+ * number of concurrent unlogged build is small.
+ *
+ * Another task of this hash is to prevent race condition during unlogged build termination.
+ * Some backend may want to evict page which backenf performing unlogged build can complete it and unlinking local files.
+ * We are using shared lock which is hold during all write operation. As far as lock is shared is doesn't prevent concurrent writes.
+ * Exclusive lock is taken by unlogged_build_end to change relation kind.
+ */
+
+typedef struct
+{
+	size_t      size;
+	uint64		hits;
+	uint64		misses;
+	uint64		pinned;
+	slock_t		mutex;
+	dlist_head	lru;			/* double linked list for LRU replacement
+								 * algorithm */
+} RelKindHashControl;
+
+static HTAB *relkind_hash;
+static LWLockId relkind_lock;
+static int	relkind_hash_size;
+static RelKindHashControl* relkind_ctl;
+static shmem_startup_hook_type prev_shmem_startup_hook = NULL;
+#if PG_VERSION_NUM >= 150000
+static shmem_request_hook_type prev_shmem_request_hook = NULL;
+static void relkind_shmem_request(void);
+#endif
+
+#define MAX_CONCURRENTLY_ACCESSED_UNLOGGED_RELS 100 /* MaxBackend? */
+
+/*
+ * Should not be smaller than MAX_CONCURRENTLY_ACCESSED_UNLOGGED_RELS.
+ * Size of a cache entry is 32 bytes. So this default will take about 2 MB,
+ * which seems reasonable.
+ */
+#define DEFAULT_RELKIND_HASH_SIZE (64 * 1024)
```

Comment:
In the worst case, the hash needs to be large enough for the case that all backends are performing an unlogged index build at the same time. Or twice that, I guess, because while performing an unlogged index build, each backend can also be trying to write out a page for another relation and hence hold one more entry in the cache pinned.

So I think the worst case is `2 * MaxBackends`. Can we use that and remove the GUC?

Reply:
200 seems to be too small for this cache. It should not only protect unlogged build, but also avoid calls of `mdexists` when pages are evicted from shared buffers. Even `go_catalog has not than 400 relations. 
Yes, I also do not like to introduce too much GUCs. But it seems to be strange to have GUC for relsize_cache and do not have it for relkind_cache.